src/routeMap.js

```diff
@@ -0,0 +1,164 @@
+
+    function routeChildrenToRoutes(children, base) {
+      return Object.entries(children).reduce(
+        (result, [k, v]) => ({ ...result, ...routeToRoutes(v, defaultHandler, overrides, base) }),
+        {}
+      )
+    }
+  }
+}
+
+function normalize({ path, meta, data, ...children }) {
+  return route({ [_]: {}, path, meta, data, ...normalizeChildren(children) })
+
+  function normalizeChildren(children) {
+    return Object.entries(children).reduce(
+      (result, [k, v]) => {
+        const route = normalize(typeof v === 'string' ? { path: v } : v)
+        return { ...result, [k]: route }
+      },
+      {}
+    )
+  }
+}
+
+// TODO: yeah: clean this up
+function route(info) {
+  const { [_]: { parentPaths = [] }, path } = info
+  return Object.assign(route, info)
+
+  function route(params) {
+    return [...parentPaths, path].reduce(
+      (base = '', path) => {
+        return (
+          base && typeof base === 'object'
+            ? Object.entries(base).reduce(
+            (result, [k, base]) => {
+              const pathValue = interpolate(path[k] || path, params)
+              return {
+                ...result,
+                [k]: base ? `${base}/${pathValue}` : pathValue
+              }
+            },
+            {}
+          )
+          : (
+            path && typeof path === 'object'
+              ? Object.entries(path).reduce(
+                (result, [k, path]) => ({
+                  ...result,
+                  [k]: base ? `${base}/${interpolate(path, params)}` : interpolate(path, params)
+                }),
+                {}
+              )
+              : base ? `${base}/${interpolate(path, params)}` : interpolate(path, params)
+          )
+        )
+      },
+      ''
+    )
+  }
+}
+
+function makePathAbsolute(path, base) {
+  return (
+    base && path ? `${base}/${pathAsString(path)}` :
+    base ? base :
+    path ? pathAsString(path) :
+    path
+  )
+}
+
+function pathAsString(path) {
+  return (
+    typeof path === 'string' ? path :
+    path ? `(?:${Object.values(path).join('|')})` :
+    path
+  )
+}
+
+function addParentPaths({ [_]: internal, path, meta, data, ...children }, parentPaths = []) {
+  return route({
+    [_]: { parentPaths }, path, meta, data,
+    ...addParentPathsToChildren(children, parentPaths.concat(path || []))
+  })
+}
+
+function addParentPathsToChildren(children, parentPaths = []) {
+  return Object.entries(children).reduce(
+    (result, [k, v]) => ({ ...result, [k]: addParentPaths(v, parentPaths)}),
+    {}
+  )
+
+}
+
+function extractPath(route) {
+  if (!route.hasOwnProperty(_)) throw new Error(`It seems the route '${JSON.stringify(route)}' is not from the route map`)
+  const { [_]: { parentPaths }, path, meta, data, ...children } = route
+  const abs = [...parentPaths, path].reduce(
+    (base, path) => makePathAbsolute(path, base),
+    ''
+  )
+  return `${abs}${Object.keys(children).length ? '/*' : ''}`
```

Comment:
While tricky, I think it should result in something like this for paths with children:

```js
/*(child1|child1/subchild1|child1/subchild2|child2/special)
```

This would allow us to detect non-existing urls at any level of the tree. And would prevent something like `/x/:y/*` to match non existing routes after the `*`.

Not sure if it is worth the added complexity though. So I think you should first finish with the current feature set.